Implement Sub-MOTD Feature

xRodzMC opened the pull request:
# What is this?
This is plugin is target to test the Sub-Motd Support for PocketMine-MP.<br>
This is on-going to be test in the future, if this repository was archived, that means it was approved to the PocketMine-MP Original Repository.
# What is Sub-MOTD?
Custom Sub-MOTD Support as on Custom MOTD in the config. This changes the interface query of the server.
[![](https://www.wikihow.com/images/thumb/7/75/Join-Servers-in-Minecraft-PE-Step-38.jpg/v4-460px-Join-Servers-in-Minecraft-PE-Step)](https://github.com/pmmp/PocketMine-MP/issues/5204)
The "My World" will change as on `server.properties` provided sub-motd. We can say its useless but if we add this feature we can say people will not going to make their own pocketmine-mp for changing `VersionInfo.php` the name. So we can use function `Network->setName();` & `Network->setLanName()` if value is missing it will returning `VersionInfo::NAME` as default. People wont make their own useless Pocketmine-MP Forks to implement this feature.
- https://github.com/pmmp/PocketMine-MP/pull/4737
- [Sub-motd (description)? (PMMP Forums)](https://forums.pmmp.io/threads/sub-motd-description.10185/)

# Changes
## API
```php
Network->setName("MOTD");
Network->setLanName("MOTD");
```
## Behaivor Changes
[![](/preview.png)](https://github.com/pmmp/PocketMine-MP/issues/5204)

## Backwards Compability
```php
$this->getServer()->getNetwork()->setName("MOTD");
$this->getServer()->getNetwork()->setLanName("MOTD");

$this->getServer()->getMotd();
$this->getServer()->getLanMotd();

$this->getServer()->getNetwork()->getName();
$this->getServer()->getNetwork()->getLanName();

$this->getServer()->getQueryInformation()->getLanServerName();
$this->getServer()->getQueryInformation()->setLanServerName("MOTD");
```

# Tests
Test the server and run with example plugin test for this feature.
- [ExamplePlugin](https://github.com/xqwtxon/SubMotd-TestPlugin)

# For Plugin Developers
Do not use `Network->internalSetName()`, this kindly mostly bug on PocketMine-MP RakLib, The Query is displaying the ipv4/ipv6. It break backwards compability and behaivor of the game. Take a example how developers can do with it.

# For Internal Developers
Please help me to fix this `internalSetName()` to protect these agains the BC Break.

xRodzMC committed:
Implement Sub-MOTD (#5204)

xRodzMC committed:
Merge branch 'stable' of https://github.com/pmmp/PocketMine-MP into stable

xRodzMC committed:
Totaly hack

xRodzMC committed:
fixes the config, if blank is provided. It would be PocketMine-MP Again

xRodzMC committed:
Opps wrong

xRodzMC committed:
no

xRodzMC committed:
Merge branch 'stable' of https://github.com/pmmp/PocketMine-MP into stable

xRodzMC committed:
Fix

xRodzMC committed:
Merge branch 'pmmp:stable' into stable

xRodzMC committed:
Backwards Compability

xRodzMC commented:
Sorry, if my commit is unverified, i havent setup yet my gpg key.

xRodzMC committed:
Missing use statement function of array_values (Dylan's bad)

NhanAZ commented on src/PocketMine.php:
```diff
@@ -245,7 +245,7 @@ function server(){
 				exit(1);
 			}
-		}
+		}*/
```
Why did you put this in the comment?

xRodzMC replied:
oh my bad

xRodzMC replied:
fixed

NhanAZ commented on tests/phpunit/block/BrewingStandTest.php:
```diff
@@ -25,6 +25,7 @@
 
 use PHPUnit\Framework\TestCase;
 use pocketmine\block\utils\BrewingStandSlot;
+use function array_values;
```
This has nothing to do with PR

xRodzMC replied:
I hate php-cs-fixer dryrun ๏_๏

xRodzMC committed:
remove comment "/ my bad

dktapps committed:
Merge branch 'stable' into stable

xRodzMC committed:
Merge branch 'stable' into stable

Endermanbugzjfc commented:
What is a “sub-MOTD”?

NhanAZ commented:
> What is a “sub-MOTD”?

https://github.com/pmmp/PocketMine-MP/issues/5204

Endermanbugzjfc commented:
> > What is a “sub-MOTD”?
> 
> 
> 
> https://github.com/pmmp/PocketMine-MP/issues/5204

Thanks.

dktapps reviewed (changes requested):
As before, the documentation available for this feature is inadequate. Most of the time, people won't even see this text, so they'll think something is broken when it doesn't appear to change anything.

matcracker commented:
Since this MOTD is only shown in the "LAN" page list, wouldn't it be better to call the parameter as "lan-motd"? At first not knowing what the sub-motd was, I was expecting a message under the current MOTD instead it only concerns the LAN screen.

xRodzMC commented:
> Since this MOTD is only shown in the "LAN" page list, wouldn't it be better to call the parameter as "lan-motd"? At first not knowing what the sub-motd was, I was expecting a message under the current MOTD instead it only concerns the LAN screen.

Good idea :)

xRodzMC committed:
Merge branch 'stable' into stable

xRodzMC committed:
Change `getSubMotd()` to `getLanMotd()` by @matcracker

Relative Fixes:
- Now you can use `Network->setLanName()` and `Network->getLanName()`
- If null the default software name will be returned in case of query bug (also lan bug)

Sorry for unverified commit. I havent configured my gpg yet.

xRodzMC committed:
php-cs-fixer

xRodzMC committed:
Fixed

xRodzMC committed:
Revert PocketMine.php

xRodzMC committed:
Please do not

xRodzMC committed:
pocketmine\network\Network::setName() invoked with 2 parameters, 1 required

xRodzMC committed:
php-cs-fixer

xRodzMC commented:
For Developers:

Please do not use `Network->internalSetName()` It breaks backwards compabilities. If you want to improve this or change this, please kindly make it at your own. If you want to help us, please kindly create your own pull requests.

xRodzMC committed:
fix?

xRodzMC committed:
Fix phpstan problems and done!

xRodzMC committed:
i hate phpstan and php-cs-fixer, it opposite together :(

xRodzMC commented:
Also, i found more links about the pull requests, such as forum and other pull requests.

dktapps committed:
Updated DevTools submodule to pmmp/DevTools@bd0fa048dae29bebe25ba76d8c58788e92418b67

xRodzMC committed:
Update dependabot.yml

xRodzMC committed:
Test

xRodzMC committed:
Update Submodule

xRodzMC committed:
Delete php

xRodzMC committed:
Merge branch 'stable' into stable

xRodzMC committed:
opps

xRodzMC commented:
My commits are so advance time, sorry but time havent configured yet.

xRodzMC committed:
Fix Conflicts

dktapps commented:
This PR has taken up way too much time and effort to be worthwhile for such a small change. It's also now venturing into BC-break territory since it adds new requirements to the NetworkInterface interface. In addition, the author seems unwilling or unable to address review comments, so I'm going to close this.

xRodzMC committed:
Fix using "php-cs-fixer"

xRodzMC committed:
Merge remote-tracking branch 'origin/stable' into stable

xRodzMC committed:
Fix

xRodzMC commented:
> This PR has taken up way too much time and effort to be worthwhile for such a small change. It's also now venturing into BC-break territory since it adds new requirements to the NetworkInterface interface. In addition, the author seems unwilling or unable to address review comments, so I'm going to close this.

Sorry, i have enough time to do work with it, Since i was busy...
